### PR TITLE
Supported nodeSelector.matchFields in scheduler.

### PR DESCRIPTION
**Pull request by k82cn**
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
part of  #61410

**Special notes for your reviewer**:

**Release note**:
```release-note
Supported nodeSelector.matchFields (node's `metadata.node`) in scheduler.
```

### CODE REVIEW

**Comment by k82cn** [comment]
@kubernetes/sig-scheduling-pr-reviews

File: pkg/scheduler/algorithm/predicates/predicates.go [hunk 1]
```diff
@@ -773,21 +775,16 @@ func PodFitsResources(pod *v1.Pod, meta algorithm.PredicateMetadata, nodeInfo *s
...
-		}
-		if nodeSelector.Matches(labels.Set(node.Labels)) {
-			return true
-		}
+	nodeFields := map[string]string{}
+	for k, f := range algorithm.NodeFieldSelectorKeys {
+		nodeFields[k] = f(node)
 	}
-	return false
+
```
Comment (ravisantoshgudimetla) [inline pkg/scheduler/algorithm/predicates/predicates.go]: nit: Remove blank line.

**Review by ravisantoshgudimetla — COMMENTED** [review]
LGTM from scheduling side. I will wait for @bsalamat to comment as he has more context on this PR.

File: pkg/scheduler/algorithm/predicates/predicates_test.go [hunk 2]
```diff
@@ -1387,11 +1388,66 @@ func TestPodFitsSelector(t *testing.T) {
 			fits: false,
 			test: "Pod with an invalid value in Affinity term won't be scheduled onto the node",
 		},
+		{
```
Comment (bsalamat) [inline pkg/scheduler/algorithm/predicates/predicates_test.go]: Could you please also add a test case with multiple selector terms? For example a case where there are multiple `MatchExpressions` and `MatchFields` terms?

File: pkg/scheduler/algorithm/types.go [hunk 3]
```diff
@@ -25,6 +25,12 @@ import (
 	"k8s.io/kubernetes/pkg/scheduler/schedulercache"
 )
 
+// NodeFieldSelectorKeys is a map that: the key are node field selector keys; the value are
```
Comment (bsalamat) [inline pkg/scheduler/algorithm/types.go]: s/the value/the values/

File: pkg/scheduler/algorithm/predicates/predicates_test.go [hunk 4]
```diff
@@ -1387,11 +1388,66 @@ func TestPodFitsSelector(t *testing.T) {
...
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			nodeName: "node_2",
+			fits:     false,
+			test:     "Pod with matchFields using In operator that not matches the existing node",
```
Comment (bsalamat) [inline pkg/scheduler/algorithm/predicates/predicates_test.go]: s/not matches/does not match/

File: pkg/scheduler/algorithm/predicates/predicates.go [hunk 5]
```diff
@@ -773,21 +775,16 @@ func PodFitsResources(pod *v1.Pod, meta algorithm.PredicateMetadata, nodeInfo *s
...
+	return v1helper.MatchNodeSelectorTerms(nodeSelectorTerms, labels.Set(node.Labels), fields.Set(nodeFields))
 }
 
 // The pod can only schedule onto nodes that satisfy requirements in both NodeAffinity and nodeSelector.
-func podMatchesNodeLabels(pod *v1.Pod, node *v1.Node) bool {
+func podMatchesNodeSelector(pod *v1.Pod, node *v1.Node) bool {
```
Comment (bsalamat) [inline pkg/scheduler/algorithm/predicates/predicates.go]: Please rename it to `podMatchesNodeSelectorAndAffinityTerms` to further distinguish it from the predicate function. Please update the comment and start it with the function name.

File: pkg/scheduler/algorithm/predicates/predicates.go [hunk 6]
```diff
@@ -773,21 +775,16 @@ func PodFitsResources(pod *v1.Pod, meta algorithm.PredicateMetadata, nodeInfo *s
...
 	// Check if node.Labels match pod.Spec.NodeSelector.
 	if len(pod.Spec.NodeSelector) > 0 {
 		selector := labels.SelectorFromSet(pod.Spec.NodeSelector)
```
Comment (bsalamat) [inline pkg/scheduler/algorithm/predicates/predicates.go]: Should we call `nodeMatchesNodeSelectorTerms` and pass the `NodeSelector` terms of the pod? Am I missing something?

Comment (k82cn) [reply pkg/scheduler/algorithm/predicates/predicates.go]: `nodeMatchesNodeSelectorTerms` only accept `NodeSelectorTerm`, but we can only get `Selector` here :(.

File: pkg/scheduler/algorithm/predicates/predicates_test.go (hunk 2 above)
Comment (k82cn) [reply pkg/scheduler/algorithm/predicates/predicates_test.go]: done :)

**Review by ravisantoshgudimetla — APPROVED** [review]
Thanks @k82cn

File: pkg/scheduler/algorithm/well_known_labels.go [hunk 7]
```diff
@@ -74,4 +78,8 @@ const (
 
 	// TaintNodeShutdown when node is shutdown in external cloud provider
 	TaintNodeShutdown = "node.cloudprovider.kubernetes.io/shutdown"
+
+	// NodeFieldSelectorKeyNodeName ('metadata.name') uses this as node field selector key
+	// when selecting node by node's name.
+	NodeFieldSelectorKeyNodeName = api.ObjectNameField
```
Comment (janetkuo) [inline pkg/scheduler/algorithm/well_known_labels.go]: This is not a label, so should not be placed in `well_known_labels.go`.

**Pull request merged**